src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -2768,15 +2755,32 @@ Handle java_lang_Throwable::get_cause_with_stack_trace(Handle throwable, TRAPS)
   Symbol* exception_name = vmSymbols::java_lang_ExceptionInInitializerError();
   Handle h_cause = Exceptions::new_exception(THREAD, exception_name, st.as_string());
 
-  // If new_exception returns a different exception while creating the exception, return null.
-  if (h_cause->klass()->name() != exception_name) {
-    log_info(class, init)("Exception thrown while saving initialization exception %s",
+  if (with_stack_trace) {
+    // If new_exception returns a different exception while creating the exception, return null.
+    if (h_cause->klass()->name() != exception_name) {
+      log_info(class, init)("Exception thrown while saving initialization exception %s",
                           h_cause->klass()->external_name());
-    return Handle();
+      return Handle();
+  }
```

Comment:
You need to do this whether `with_stack_trace` or not, otherwise you will install the exception from `new_exception` as the initialization error for this class, when in fact it was not the initialization error at all.

Reply:
I would suggest a correction in this place. 
Just because if Java is in a state, when it throws an SOE at JavaCalls::call_virtual(), the most probably that the `Exceptions::new_exception()`, which actually wraps up a call to Java, also throws an SOE. 
If I return an empty Handle(), any clues on the cause will be missed from a stacktrace, even one of my testcases shows the following results: 

```
Exception in thread "main" java.lang.NoClassDefFoundError: Could not initialize class java.lang.invoke.MethodHandleImpl$AsVarargsCollector
	at java.sql/java.sql.DriverManager.ensureDriversInitialized(DriverManager.java:628)
	at java.sql/java.sql.DriverManager.getDrivers(DriverManager.java:427)
	at jdbc.repro.DriverA.<init>(DriverA.java:22)
	at jdbc.repro.DriverRepro.main(DriverRepro.java:12)
Caused by: java.lang.NoClassDefFoundError: Could not initialize class java.lang.invoke.MethodHandleImpl$AsVarargsCollector
	... 4 more
```